# Basics of Python and Data Analysis
## Study of Export of Chinese bus market

Below is a study of Chinese bus export based on the national Statistic Bureau Data for the period of January to May 2021. The study is done to master the theoretical material learned at Yandex Praktikum in Basics of Python and Data Analysis module. The plan of the study is following:

<a href='#1'>1. Data loading </a>

<a href='#2'>2. Data pre-processing </a>

<a href='#3'>3. Data analysis </a>

<a href='#4'>4. Conclusion </a>

<a id='1'></a> 
 ### 1. Data loading

In [1]:
#loading necessary libraries
import pandas as pd
import gspread
from oauth2client.service_account import ServiceAccountCredentials
from pandas.io.json import json_normalize
from unicodedata import normalize

In [2]:
## Connect to the service account
scope = ['https://spreadsheets.google.com/feeds']
credentials = ServiceAccountCredentials.from_json_keyfile_name('/Users/anaitagadzhanyan/Desktop/Практикум/extreme-minutia-321117-677aeb6efe7b.json', scope)
gc = gspread.authorize(credentials)

In [3]:
##Get data sheet from Google Drive
spreadsheet_key = '1I8050Yy6l3K4Z0EQEaf1OS85S-onnQRvY5jaiWoAd0E'
book = gc.open_by_key(spreadsheet_key)
worksheet = book.worksheet("原始数据")
table = worksheet.get_all_values()

In [4]:
##Convert table data into a dataframe
df = pd.DataFrame(table[1:], columns=table[0])
display(df.head())

日期 进出口      商品编码   车型   座位数 按发动机 二手车  海关编码描述  \
0  202101\t\t  出口  87021091  大中客  30+座   柴油  新车  30+座柴油   
1  202102\t\t  出口  87021091  大中客  30+座   柴油  新车  30+座柴油   
2  202102\t\t  出口  87021091  大中客  30+座   柴油  新车  30+座柴油   
3  202101\t\t  出口  87021091  大中客  30+座   柴油  新车  30+座柴油   
4  202102\t\t  出口  87021091  大中客  30+座   柴油  新车  30+座柴油   

                                      商品名称   目的国  ...  原产地  大区 成交方式  \
0  仅装有压燃式活塞内燃发动机的大型客车,≥30座                    文莱  ...  江苏省  华东  FOB   
1  仅装有压燃式活塞内燃发动机的大型客车,≥30座                    缅甸  ...  山东省  华东  FOB   
2              仅装有压燃式活塞内燃发动机的大型客车,≥30座\t\t    缅甸  ...  河南省  华中  FOB   
3              仅装有压燃式活塞内燃发动机的大型客车,≥30座\t\t  中国香港  ...  广东省  华南  FOB   
4              仅装有压燃式活塞内燃发动机的大型客车,≥30座\t\t  中国香港  ...  广东省  华南  FOB   

           单价         美元  币制 法定数量   法定单位    区域  所属大区  
0   68500,00     137 000  美元    2  辆\t\t  亚太一区  亚太大区  
1   77487,00     774 870  美元   10  辆\t\t  亚太二区  亚太大区  
2  101690,00   2 033 800  美元   20  辆\t\t  亚太二区  亚太大区  
3  109412,00     109 412  美元    1  辆\t\t  亚太三区  亚太大区  
4  136998,00     136 998  美元    1  辆\t\t  亚太三区  亚太大区  

[5 rows x 24 columns]

In [5]:
#get basic information obout the data
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 787 entries, 0 to 786
Data columns (total 24 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   日期      787 non-null    object
 1   进出口     787 non-null    object
 2   商品编码    787 non-null    object
 3   车型      787 non-null    object
 4   座位数     787 non-null    object
 5   按发动机    787 non-null    object
 6   二手车     787 non-null    object
 7   海关编码描述  787 non-null    object
 8   商品名称    787 non-null    object
 9   目的国     787 non-null    object
 10  产销经济区   787 non-null    object
 11  产销州     787 non-null    object
 12  产销地里区   787 non-null    object
 13  贸易方式    787 non-null    object
 14  原产地     787 non-null    object
 15  大区      787 non-null    object
 16  成交方式    787 non-null    object
 17  单价      787 non-null    object
 18  美元      787 non-null    object
 19  币制      787 non-null    object
 20  法定数量    787 non-null    object
 21  法定单位    787 non-null    object
 22  区域      787 non-null    ob

We have got the `df` of 787 entries of 24 parameters. The number of observations is the same for every parameter, meaning that we don't have any missing values.

<a id='2'></a> 
### 2. Data pre-processing

In [6]:
#removing columns that are not necessary

df.drop(['进出口', '海关编码描述','商品名称','产销地里区','法定单位'], axis=1, inplace=True) 


#changing column names
df.set_axis(['date', 'hs_code', 'model', 'seat_no', 'engine', 'used',
            'country','area','continent','conditions',
            'origin','province','inctm','price_rmb', 'price_usd','currency','qty',
            'department','depratment_big'], axis='columns', inplace=True)

In [7]:
print(df.info())
display(df.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 787 entries, 0 to 786
Data columns (total 19 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   date            787 non-null    object
 1   hs_code         787 non-null    object
 2   model           787 non-null    object
 3   seat_no         787 non-null    object
 4   engine          787 non-null    object
 5   used            787 non-null    object
 6   country         787 non-null    object
 7   area            787 non-null    object
 8   continent       787 non-null    object
 9   conditions      787 non-null    object
 10  origin          787 non-null    object
 11  province        787 non-null    object
 12  inctm           787 non-null    object
 13  price_rmb       787 non-null    object
 14  price_usd       787 non-null    object
 15  currency        787 non-null    object
 16  qty             787 non-null    object
 17  department      787 non-null    object
 18  depratment

date   hs_code model seat_no engine used country area continent  \
0  202101\t\t  87021091   大中客    30+座     柴油   新车      文莱  东南亚        亚洲   
1  202102\t\t  87021091   大中客    30+座     柴油   新车      缅甸  东南亚        亚洲   
2  202102\t\t  87021091   大中客    30+座     柴油   新车      缅甸  东南亚        亚洲   
3  202101\t\t  87021091   大中客    30+座     柴油   新车    中国香港   亚太        亚洲   
4  202102\t\t  87021091   大中客    30+座     柴油   新车    中国香港   亚太        亚洲   

   conditions origin province inctm   price_rmb  price_usd currency qty  \
0    一般贸易\t\t    江苏省       华东   FOB   68500,00     137 000       美元   2   
1    一般贸易\t\t    山东省       华东   FOB   77487,00     774 870       美元  10   
2    一般贸易\t\t    河南省       华中   FOB  101690,00   2 033 800       美元  20   
3  来料加工贸易\t\t    广东省       华南   FOB  109412,00     109 412       美元   1   
4  来料加工贸易\t\t    广东省       华南   FOB  136998,00     136 998       美元   1   

  department depratment_big  
0       亚太一区           亚太大区  
1       亚太二区           亚太大区  
2       亚太二区           亚太大区  
3       亚太三区           亚太大区  
4       亚太三区           亚太大区

In [8]:
#checking if there are missing values
df.isnull().sum()

date              0
hs_code           0
model             0
seat_no           0
engine            0
used              0
country           0
area              0
continent         0
conditions        0
origin            0
province          0
inctm             0
price_rmb         0
price_usd         0
currency          0
qty               0
department        0
depratment_big    0
dtype: int64

In [9]:
#checking if there are duplicates
df.duplicated().sum()

0

In [10]:
#set numeric value

df.loc[:,['qty']] = df.loc[:,['qty']].astype('int')

In [11]:
#function to replace used parameter with bullet values
def used_true(df, used_value):
    result = 0
    df['used'] = df['used'].replace(used_value, True)
    result = df[df['used'] == used_value]['used'].count()
    return result

def used_false(df, used_value):
    result = 0
    df['used'] = df['used'].replace(used_value, False)
    result = df[df['used'] == used_value]['used'].count()
    return result

used_true(df, '二手车')
used_false(df, '新车')

display(df.head())

date   hs_code model seat_no engine   used country area continent  \
0  202101\t\t  87021091   大中客    30+座     柴油  False      文莱  东南亚        亚洲   
1  202102\t\t  87021091   大中客    30+座     柴油  False      缅甸  东南亚        亚洲   
2  202102\t\t  87021091   大中客    30+座     柴油  False      缅甸  东南亚        亚洲   
3  202101\t\t  87021091   大中客    30+座     柴油  False    中国香港   亚太        亚洲   
4  202102\t\t  87021091   大中客    30+座     柴油  False    中国香港   亚太        亚洲   

   conditions origin province inctm   price_rmb  price_usd currency qty  \
0    一般贸易\t\t    江苏省       华东   FOB   68500,00     137 000       美元   2   
1    一般贸易\t\t    山东省       华东   FOB   77487,00     774 870       美元  10   
2    一般贸易\t\t    河南省       华中   FOB  101690,00   2 033 800       美元  20   
3  来料加工贸易\t\t    广东省       华南   FOB  109412,00     109 412       美元   1   
4  来料加工贸易\t\t    广东省       华南   FOB  136998,00     136 998       美元   1   

  department depratment_big  
0       亚太一区           亚太大区  
1       亚太二区           亚太大区  
2       亚太二区           亚太大区  
3       亚太三区           亚太大区  
4       亚太三区           亚太大区

In [12]:
#coverting usd_price values to integers

def norm_func(df, column):
    for i in range(len(df)):
        df[column][i] = normalize('NFKD', df[column][i])
    for i in range(len(df)):
        df[column][i] = df[column][i].replace(' ','')
    for i in range(len(df)):
        df[column][i] = df[column][i].replace(',','.')
    for i in range(len(df)):
        df[column][i] = int(df[column][i])

norm_func(df,'price_usd')

<ipython-input-12-a9dbbf523dc4>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[column][i] = normalize('NFKD', df[column][i])
<ipython-input-12-a9dbbf523dc4>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[column][i] = df[column][i].replace(' ','')
<ipython-input-12-a9dbbf523dc4>:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[column][i] = df[column][i].replace(',','.')
<ipython-input-12-a9dbbf523dc4>:11: SettingWithCopyWar

We have removed the columns `进出口, 海关编码描述, 商品名称, 产销地里区, 法定单位` that we won't need in the further analysis and gave more convenient names to the remaining columns. We haven't found any missing values or duplicates. We wrote two functions to replace the values of the column `used` with bullet values. We also found our that the `usd_price` column values are not normalized and are not suitable to use for the further quantitively  analysis. We have normalized the values and converted them to integers. 

<a id='3'></a> 
### 3. Data analysis

In [13]:
#calculate the sum for every order
df['sum_usd'] = df['price_usd'] * df['qty']

#calculate parameters by areas
print('Total export volumes by areas:')
display(df.groupby('area')['sum_usd'].sum().sort_values(ascending = False).head())
print('Total export quantities by areas:')
display(df.groupby('area')['qty'].sum().sort_values(ascending = False).head())

Total export volumes by areas:


area
南亚     15280889910
南美洲    13666245696
东非     10824855936
西亚      3516009441
西欧      2214936713
Name: sum_usd, dtype: int64

Total export quantities by areas:


area
东非     4367
南美洲    3683
西亚     1645
东南亚    1424
南亚     1187
Name: qty, dtype: int64

In [14]:
#function for parameters of a country

def parameter_country(df, country, parameter):
    country_list = df.loc[(df['country'] == country)]
    country_list_count = country_list[parameter].sum()
    return country_list_count

print('Number of buses exported to Myanmar:')
parameter_country(df, '缅甸','qty')

Number of buses exported to Myanmar:


63

In [15]:
print('Sum value of buses exported to Myanmar:')
parameter_country(df, '缅甸','sum_usd')

Sum value of buses exported to Myanmar:


53058875

In [16]:
#making a table
data = [['Myanmar', 63, 53058875]]
columns = ['Country','Number of buses', 'Sum of value']
table = pd.DataFrame(data = data, columns = columns)
print(table)

   Country  Number of buses  Sum of value
0  Myanmar               63      53058875


<a id='4'></a> 
### 4. Conclusion

We have uploaded the data from google drive directory, converted into a data frame and conducted data pre-processing. We have removed the columns that we are not going to use in the further analysis, checked if there are any duplicates or missing values and converted to columns data types for further analysis.

We have calculated the sum value of each row and grouped data by areas of export. We have found out that South Asia had the highest export volume in the period of January to May 2021 with 15 280 889 910 USD value. However, South Asia is only on the 5th place by quantity of exported buses, with 1187 buses. The highest quantity of exported buses is East Africa, with the export volume of  1 0824 855 936 USD and 4367 buses. That gives us grounds to say that the buses exported to East Africa on average were cheaper then those exported to South Asia. We have also wrote a function find the required value for a certain country. 

In the further analysis it is worth examining the data over the time (that will require manipulation with the date column) and plotting pie charts and other graphical representation of our data.